DOCS-1520: Document App Client new methods (#2359)
viamrobotics/docs: docs/build/program/apis/cloud.md

{{% /tab %}}
{{< /tabs >}}

### CreateOrganizationInvite

{{< tabs >}}
{{% tab name="Python" %}}

> Create an organization invite and send it by email.

Reviewer: ```suggestion
Create an {{< glossary_tooltip term_id="organization" text="organization" >}} invite and send it by email.
```
Consider adding glossary defs to Viam-specific terms in every description

**Parameters:**

- `email` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The email address to send the invite to.

> - `authorizations` [(Optional[List[viam.proto.app.Authorization]])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.Authorization): Specifications of the
>   authorizations to include in the invite.

Reviewer: ```suggestion
- `authorizations` [(Optional[List[viam.proto.app.Authorization]])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.Authorization): Specifications of the authorizations to include in the invite.
```
One line I think?

Also, can you add an example or two of what a "specification" would be? Or is that not practical and the user should just click the link for more info (fine if that's the case)?

Author: It just means those objects. I'll remove the specifications phrasing

  If not provided, full owner permissions will be granted.

**Returns:**

- None.

**Raises:**

- `GRPCError`: If an invalid email is provided, or if the user is already a member of the org.

```python {class="line-numbers linkable-line-numbers"}
await cloud.create_organization_invite("youremail@email.com")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.create_organization_invite).

{{% /tab %}}
{{< /tabs >}}

### DeleteOrganizationMember

{{< tabs >}}
{{% tab name="Python" %}}

> Remove a member from the organization.

Author: @stuqdog Is this accurate?

> Remove a member from the organization.

Reviewer 2: That is correct, at least for now! Everything is subject to change once we have user-level auth in the python SDK.

> Remove a member from the organization.

Author: @stuqdog Thanks!

**Parameters:**

- `user_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the user to remove.

**Returns:**

- None.

```python {class="line-numbers linkable-line-numbers"}
member_list, invite_list = await cloud.list_organization_members()
first_user_id = member_list[0].user_id

await cloud.delete_organization_member(first_user_id)
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.delete_organization_member).

{{% /tab %}}
{{< /tabs >}}

### DeleteOrganizationInvite

{{< tabs >}}
{{% tab name="Python" %}}

> Delete a pending organization invite.

Reviewer: [opt] Would it make sense to say
```suggestion
Delete a pending organization invite to the organization you are currently authenticated to.
```

**Parameters:**

- `email` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The email address the pending invite was sent to.

**Returns:**

- None.

**Raises:**

- `GRPCError`: If no pending invite is associated with the provided email address.

```python {class="line-numbers linkable-line-numbers"}
await cloud.delete_organization_invite("youremail@email.com")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.delete_organization_invite).

{{% /tab %}}
{{< /tabs >}}

### ResendOrganizationInvite

{{< tabs >}}
{{% tab name="Python" %}}

Resend a pending organization invite email.

**Parameters:**

- `email` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The email address associated with the invite.

**Returns:**

- None.

**Raises:**

> - `GRPCError`: If no pending invite is associated with the provided email address.

Reviewer: ```suggestion
- `GRPCError`: This error is raised if no pending invite is associated with the provided email address.
```

```python {class="line-numbers linkable-line-numbers"}
await cloud.resend_organization_invite("youremail@email.com")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.resend_organization_invite).

{{% /tab %}}
{{< /tabs >}}

### GetRoverRentalRobots

{{< tabs >}}
{{% tab name="Python" %}}

> Return a list of rover rental robots within an org.

Reviewer: ```suggestion
Return a list of rover rental robots within an {{< glossary_tooltip term_id="organization" text="organization" >}}.
```

**Parameters:**

- None.

**Returns:**

- [(List[viam.proto.app.RoverRentalRobot])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.RoverRentalRobot): The list of rover rental robots.

```python {class="line-numbers linkable-line-numbers"}
rental_robots = await cloud.get_rover_rental_robots()
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.get_rover_rental_robots).

{{% /tab %}}
{{< /tabs >}}

### CheckPermissions

{{< tabs >}}
{{% tab name="Python" %}}

> Check validity of a list of permissions.

Reviewer: Can you include a quick description of what validity means?

Reviewer: How about linking to the permissions docs as well for context?

> Check validity of a list of permissions.

Author: I assume this means they are authorized for a resource? @stuqdog Can you confirm

Author: Added that below

Reviewer 2: Yeah so `check_permissions` takes a list of permissions and returns a filtered list of which ones are actual, so e.g., if you have permission to `control_robot` and you pass a `control_robot` permission then you will get that permission back. If you _don't_ have `control_robot` permission and you pass it as your argument, you will get nothing back.

**Parameters:**

> - `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions to validate. For example, `“read_organization”` or `“control_robot”`.

Reviewer: ```suggestion
- `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions to validate.
  For example, `“read_organization”` or `“control_robot”`.
```

> - `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions to validate.

Author: ```suggestion
- `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions to validate as authorized for a resource.
```

> - `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): A list of the resource permissions arguments to validate as authorized.

Reviewer: I find this hard to parse. How about 
```suggestion
- `permissions` [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): A list containing the resource type and ID, followed by the resource permissions to check.
```

>   For example, `“read_organization”` or `“control_robot”`.

Reviewer: ```suggestion
```
Since this example doesn't contain the resource type and ID, maybe take it out and instead let them refer to the example below, and add a comment to that example.

>   See [GitHub](https://github.com/viamrobotics/app/blob/main/auth/models/models.go#L190) for the full list of permissions.

Author: @JessamyT Do you have any idea why this GitHub link keeps on failing htmltest? Idk what is wrong

Reviewer: Hmm I'm not sure--have you re-run it multiple times?

Reviewer 3: Because it's not public code. You should not link to app code - is there something else you can link to?

>   See [GitHub](https://github.com/viamrobotics/app/blob/7ddb29e4dcb8cbca6c76c3acaeeb19c544028bd2/auth/models/models.go#L190) for the full list of permissions.

Author: ```suggestion
  See [GitHub](https://github.com/viamrobotics/app/blob/main/auth/models/models.go#L190) for the full list of permissions.
```

**Returns:**

- [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions argument, with invalid permissions filtered out.

**Raises:**

> - `GRPCError`: If the list of permissions to validate is empty.

Reviewer: ```suggestion
- `GRPCError`: This error is raised if the list of permissions to validate is empty.
```

> - [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions argument, with invalid permissions filtered out.

Reviewer: Hmm I'm realizing I'm not sure what this method does. @stuqdog ?
```suggestion
- [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions argument, with invalid permissions filtered out.
For example, if you pass a location with its ID, followed by permissions `"read_fragment"` and `"write_fragment"`, and the location can read but not write fragments in the organization you are currently authenticated to, the method will return `["read_fragment"]`.
```
or is it
```suggestion
- [(List[viam.proto.app.AuthorizedPermissions])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.AuthorizedPermissions): The permissions argument, with invalid permissions filtered out.
For example, if you pass a location with its ID, followed by permissions `"read_fragment"` and `"write_fragment"`, and the organization you are currently authenticated to can read but not write fragments in the specified location, the method will return `["read_fragment"]`.
```
or none of the above? Sorry if these are bad examples--not sure how it works!

Reviewer 2: I _believe_ it's the former. You pass some sort of ID (org, user, location, etc.) and then it returns which permissions that ID has access to. I suspect fleet knows more. @Kschappacher you were the last one to touch the `AuthorizedPermissions` type in app.proto (admittedly many months ago), can you confirm?

Reviewer 4: Yupp that is correct. You can pass in any kind of resource id along with the permission you want to check and it will return subset of permissions the resource  actually has access to.

```python {class="line-numbers linkable-line-numbers"}
from viam.proto.app import AuthorizedPermissions

> permissions = [AuthorizedPermissions(resource_type="organization",

Reviewer: ```suggestion
# Check whether organization of ID "organization-identifier123" has permissions to control robots and/or read robot logs within the organization you are currently authenticated to
permissions = [AuthorizedPermissions(resource_type="organization",
```
Or whatever is true...I think this method is confusing and it'd be good to clarify

Reviewer: ```suggestion
# Check whether the entity you're currently authenticated to has permission to control and/or read logs from robots in the organization-identifier123 org
permissions = [AuthorizedPermissions(resource_type="organization",
```
Based on offline convo I think it's actually this?

Author: Awesome, thanks!

                                     resource_id="organization-identifier123",

>                                      permissions=["control_robot"])]

Reviewer: ```suggestion
                                     permissions=["control_robot", "read_robot_logs"])]
```
To make it clear they can pass a list

filtered_permissions = await cloud.check_permissions(permissions)
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.check_permissions).

{{% /tab %}}
{{< /tabs >}}

### CreateKey

{{< tabs >}}
{{% tab name="Python" %}}

> Create a new API key.

Reviewer: Would it make sense to link to a relevant part of the docs about API keys?

Author: I can't find one besides the CLI. I think the update to RBAC I'm working on could be a good place to link to in the future.

**Parameters:**

- `authorizations` [(List[APIKeyAuthorization])](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.APIKeyAuthorization): A list of authorizations to associate with the key.
- `name` [(Optional[str])](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): A name for the key. If None, defaults to the current timestamp.

**Returns:**

- [(Tuple[str, str])](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The api key and api key ID.

**Raises:**

> - `GRPCError`: If the authorizations list is empty.

Reviewer: ```suggestion
- `GRPCError`: This error is raised if the authorizations list is empty.
```

```python {class="line-numbers linkable-line-numbers"}
from viam.app.app_client import APIKeyAuthorization

auth = APIKeyAuthorization(
  role="owner",
  resource_type="robot",
  resource_id="your-robot-id123"
)

api_key, api_key_id = cloud.create_key([auth], "my_key")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.create_key).

{{% /tab %}}
{{< /tabs >}}

### CreateKeyFromExistingKeyAuthorizations

{{< tabs >}}
{{% tab name="Python" %}}

Create a new API key with an existing key’s authorizations.

**Parameters:**

- `id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the API key to duplicate authorizations from.

**Returns:**

> - [(Tuple[str, str])](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The new api key and api key ID.

Reviewer: ```suggestion
- [(Tuple[str, str])](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The new API key and API key ID.
```

```python {class="line-numbers linkable-line-numbers"}
api_key, api_key_id = cloud.create_key_from_existing_key_authorizations(
  id="INSERT YOUR API KEY ID")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.create_key_from_existing_key_authorizations).

{{% /tab %}}
{{< /tabs >}}

### ListKeys

{{< tabs >}}
{{% tab name="Python" %}}

> List all keys for the currently-authed-to org.

Reviewer: ```suggestion
List all keys for the {{< glossary_tooltip term_id="organization" text="organization" >}} that you are currently authenticated to.
```

- `GRPCError`: This error is raised if no pending invite is associated with the provided email address.

```python {class="line-numbers linkable-line-numbers"}
await cloud.resend_organization_invite("youremail@email.com")
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.resend_organization_invite).

{{% /tab %}}
{{< /tabs >}}

### GetRoverRentalRobots

{{< tabs >}}
{{% tab name="Python" %}}

Return a list of rover rental robots within an {{< glossary_tooltip term_id="organization" text="organization" >}}.

**Parameters:**

- None.

**Returns:**

- [(List[viam.proto.app.RoverRentalRobot])](https://python.viam.dev/autoapi/viam/proto/app/index.html#viam.proto.app.RoverRentalRobot): The list of rover rental robots.

```python {class="line-numbers linkable-line-numbers"}
rental_robots = await cloud.get_rover_rental_robots()
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/app_client/index.html#viam.app.app_client.AppClient.get_rover_rental_robots).

{{% /tab %}}
{{< /tabs >}}

### CheckPermissions

{{< tabs >}}
{{% tab name="Python" %}}

> Check if the organization, location, or robot your `ViamClient` is authenticated to is permitted to perform some action or set of actions on the resource you pass to the method.

Reviewer: Love this wording--much more descriptive, nice.

{{% /tab %}}
{{< /tabs >}}

Valid arguments for permissions are as follows:

{{% expand "Click to see permissions strings available" %}}

```python
"read_organization"
"write_organization"

"read_fragment"
"write_fragment"

"read_location"
"write_location"

"read_location_secret"
"read_robot_secret"

"read_robot"
"read_robot_config"
"read_robot_logs"
"write_robot"
"control_robot"

"read_organization_data_management"
"read_location_data_management"
"read_robot_data_management"
"write_organization_data_management"
"write_location_data_management"
"write_robot_data_management"

"read_robot_history"

"read_mapping_sessions"
"create_maps"

"write_private_registry_item"
"write_public_registry_item"
"read_private_registry_item"

"train_models"

"read_packages"
"write_packages"
"delete_packages"

"configure_database_user"
"get_database_connection"

"create_dataset"
"list_dataset"
"rename_dataset"
"delete_dataset"
```

{{% /expand %}}

For more information about managing permissions, see [Role-Based Access Control](/fleet/rbac/#permissions).

### CreateKey

{{< tabs >}}
{{% tab name="Python" %}}

> Create a new API key.

Author: ```suggestion
Create a new [API key](/fleet/rbac/#api-keys).
```
